src/golang/lib/collections/execution_environment/standard.go

```diff
@@ -161,3 +162,55 @@ func (w *standardWriterImpl) DeleteExecutionEnvironments(
 	args := stmt_preparers.CastIdsListToInterfaceList(ids)
 	return db.Execute(ctx, deleteStmt, args...)
 }
+
+func (r *standardReaderImpl) GetUnusedExecutionEnvironments(
+	ctx context.Context,
+	db database.Database,
+) ([]DBExecutionEnvironment, error) {
+	// Note that we use `OperatorToArtifactType` as the filtering condition because an operator
+	// is guaranteed to generate at least one artifact, so this filter is guaranteed to capture
+	// all operators involved in a workflow DAG.
+	query := fmt.Sprintf(`
+	WITH latest_workflow_dag AS
+	(
+		SELECT 
+			workflow_dag.id 
+		FROM
+			workflow_dag 
+		WHERE 
+			created_at IN (
+				SELECT 
+					MAX(workflow_dag.created_at) 
+				FROM 
+					workflow, workflow_dag 
+				WHERE 
+					workflow.id = workflow_dag.workflow_id 
+				GROUP BY 
+					workflow.id
+			);
```

Comment:
@cw75 which version of sqlite3 you are using? I had to remove this `;` to make the code working in my setting